app/models/relay_state.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'sentry_logging'
+
+class RelayState
+  extend ActiveModel::Naming
+  include ActiveModel::Validations
+  include SentryLogging
+
+  attr_accessor :relay_enum, :url
+
+  RELAY_KEYS = Settings.saml.relays.keys.map(&:to_s).freeze
+  LOGIN_URLS = (Settings.saml.relays.to_h.values + [Settings.review_instance_slug]).compact.freeze
+  LOGOUT_URLS = (Settings.saml.logout_relays.to_h.values + [Settings.saml.logout_relay]).compact.freeze
+  ALL_RELAY_URLS = (LOGIN_URLS + LOGOUT_URLS).freeze
+
+  validates :relay_enum, allow_blank: true, inclusion: { in: RELAY_KEYS, message: '[%<value>s] not a valid relay enum' }
+  validates :url, allow_blank: true, inclusion: { in: ALL_RELAY_URLS, message: '[%<value>s] not a valid relay url' }
+  # TODO: Custom validator that BOTH are not blank
+
+  def initialize(relay_enum: nil, url: nil)
+    @relay_enum = relay_enum
+    @url = url
+
+    unless valid?
+      log_message_to_sentry(
+        'Invalid SAML RelayState!', :error,
+        error_messages: errors.messages, url_whitelist: ALL_RELAY_URLS, enum_whitelist: RELAY_KEYS
+      )
+    end
+  end
+
+  def login_url
+    return default_login_url if @relay_enum.blank? && @url.blank?
+    return Settings.saml.relays.vetsgov if @url.present? && @url == Settings.review_instance_slug
+    get_url(LOGIN_URLS + [], :relays)
+  end
+
+  def logout_url
+    return default_logout_url if @relay_enum.blank? && @url.blank?
+    return Settings.saml.logout_relay if @url.present? && @url == Settings.review_instance_slug
+    get_url(LOGOUT_URLS, :logout_relays)
+  end
+
+  private
+
+  def get_url(valid_urls, type)
+    return get_default(type) unless valid?
+    return get_default(type) if @url.present? && valid_urls.include?(@url) == false
+
+    if @url.present?
+      return @url
+    elsif Settings.saml[type][@relay_enum].present?
+      return Settings.saml[type][@relay_enum]
+    else
+      return get_default(type)
+    end
+  end
+
+  def get_default(type = :relays)
+    type == :logout_relays ? default_logout_url : default_login_url
+  end
+
+  def default_login_url
+    Settings.saml.relays&.vetsgov || Settings.saml.relay
+  end
+
+  def default_logout_url
+    Settings.saml.logout_relays&.vetsgov || Settings.saml.logout_relay
+  end
+end
```

Comment:
can you check and make sure that test coverage on this file is 100%. Actually, that might not be 100% truthful either because I'm pretty sure SimpleCov will treat something like a ternary operation as tested, even if it only tests the truth operation. In other words, you won't necessarily have complete coverage for something when there are two expressions. Here is an example

```
return get_default(type) if @url.present? && valid_urls.include?(@url) == false
```

Simplecov would say that this line has coverage even if you've only tested `@url.present? && valid_urls.include?(@url) == false` being a truthy expression. It won't necessarily test the false conditions. That being said, this is a bad example because if it is false and the rest of the expression is tested, then you satisfy the test for false. But you see where I am going with this... This is an important new addition to login and we won't have time to test all the edge cases. We need to make sure every edge case is tested.

Reply:
Report shows 100% coverage and I can confirm that this use-case is covered.  The use case is when a valid login url is supplied, but someone called `.logout_url`.

```ruby
@relay_state = RelayState.new(url: <VALID LOGIN URL>)
@relay_state.logout_url
```

I think this guard clause is not very readable and could/should be re-written as logout & login-specific model validations.  That way, the prior line of, `return default unless valid?` will be the only necessary guard clause.